src/main/java/de/tum/cit/aet/artemis/programming/service/localci/LocalCIEventListenerService.java

```diff
@@ -0,0 +1,182 @@
+package de.tum.cit.aet.artemis.programming.service.localci;
+
+import java.time.ZonedDateTime;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+import jakarta.annotation.PostConstruct;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.context.annotation.Profile;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Service;
+
+import com.hazelcast.collection.IQueue;
+import com.hazelcast.collection.ItemEvent;
+import com.hazelcast.collection.ItemListener;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.map.IMap;
+import com.hazelcast.map.listener.EntryAddedListener;
+import com.hazelcast.map.listener.EntryRemovedListener;
+import com.hazelcast.map.listener.EntryUpdatedListener;
+
+import de.tum.cit.aet.artemis.buildagent.dto.BuildAgentInformation;
+import de.tum.cit.aet.artemis.buildagent.dto.BuildJobQueueItem;
+import de.tum.cit.aet.artemis.programming.domain.build.BuildJob;
+import de.tum.cit.aet.artemis.programming.domain.build.BuildStatus;
+import de.tum.cit.aet.artemis.programming.dto.SubmissionProcessingDTO;
+import de.tum.cit.aet.artemis.programming.repository.BuildJobRepository;
+import de.tum.cit.aet.artemis.programming.service.ProgrammingMessagingService;
+
+@Service
+@Profile("localci & scheduling")
+public class LocalCIEventListenerService {
+
+    private static final Logger log = LoggerFactory.getLogger(LocalCIEventListenerService.class);
+
+    private final HazelcastInstance hazelcastInstance;
+
+    private final LocalCIQueueWebsocketService localCIQueueWebsocketService;
+
+    private final BuildJobRepository buildJobRepository;
+
+    private final SharedQueueManagementService sharedQueueManagementService;
+
+    private final ProgrammingMessagingService programmingMessagingService;
+
+    public LocalCIEventListenerService(@Qualifier("hazelcastInstance") HazelcastInstance hazelcastInstance, LocalCIQueueWebsocketService localCIQueueWebsocketService,
+            BuildJobRepository buildJobRepository, SharedQueueManagementService sharedQueueManagementService, ProgrammingMessagingService programmingMessagingService) {
+        this.hazelcastInstance = hazelcastInstance;
+        this.localCIQueueWebsocketService = localCIQueueWebsocketService;
+        this.buildJobRepository = buildJobRepository;
+        this.sharedQueueManagementService = sharedQueueManagementService;
+        this.programmingMessagingService = programmingMessagingService;
+    }
+
+    /**
+     * Add listeners for build job, build agent changes.
+     */
+    @PostConstruct
```

Comment:
this should not be done in `@PostConstruct`, because it can make the application start much slower, please use 

```
    @EventListener(ApplicationReadyEvent.class)
    public void init() {
```